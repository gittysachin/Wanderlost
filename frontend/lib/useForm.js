import { useState } from 'react';

const uploadFile = async ({ value, inputs, updateInputs }) => {
  console.log('Uploading file');
  const data = new FormData();
  data.append('file', value);
  data.append('upload_preset', 'sickfits');

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/davyhausser/image/upload',
    {
      method: 'POST',
      body: data,
    },
  );
  const file = await res.json();
  console.log(file);
  updateInputs({
    ...inputs,
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  })
};

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = e.target.files;
      uploadFile({ value, inputs, updateInputs });
    }
    updateInputs({
      ...inputs,
      [name]: e.target.value,
    });
  }

  function resetForm() {
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm
  };
}