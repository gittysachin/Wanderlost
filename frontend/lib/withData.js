import withApollo from 'next-with-apollo';
import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
      request: (operation) => {
        operation.setContext({
          fetchOptions: {
            credentials: 'include',
          },
          headers,
        });
      },
    }),
    resolvers: {
      Mutation: {
        toggleCart(_, variables, { cache }) {
          // read the cartOpen value from the cache
          const { cartOpen } = cache.readQuery({
            query: LOCAL_STATE_QUERY,
          });
          // Write the cart state to the opposite
          const data = {
            data: { cartOpen: !cartOpen },
          };
          cache.writeData(data);
          return data;
        },
      },
    },
  });
}

export default withApollo(createClient);
