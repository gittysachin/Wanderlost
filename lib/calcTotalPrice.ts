import { GetCurrentUser_user_cart } from '../components/__generated__/GetCurrentUser'

export default function calcTotalPrice(
  cart: Array<GetCurrentUser_user_cart>
): number {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally
    return tally + cartItem.quantity * cartItem.item.price
  }, 0)
}
