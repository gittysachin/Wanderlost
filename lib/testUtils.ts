import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'
import { GetItem_item } from '../components/__generated__/GetItem'
import { GetOrder_order, GetOrder_order_items } from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides: Partial<GetCurrentUser_user_cart>): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export { LocalStorageMock, fakeItem, fakeUser, fakeCartItem, fakeOrder, fakeOrderItem }
