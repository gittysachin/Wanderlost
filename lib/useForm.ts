import { useState, ChangeEvent, Dispatch, SetStateAction } from 'react'

type Inputs = Record<string, string>
type Event = ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
type Form = {
  inputs: Inputs
  setInputs: Dispatch<SetStateAction<Inputs>>
  handleChange: (event: Event) => void
  resetForm: () => void
  clearForm: () => void
}

async function uploadFile({ value, inputs, setInputs }) {
  const data = new FormData()
  data.append('file', value)
  data.append('upload_preset', 'sickfits')

  const res = await fetch('https://api.cloudinary.com/v1_1/davyhausser/image/upload', {
    method: 'POST',
    body: data,
  })
  const file = await res.json()
  setInputs({
    ...inputs,
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  })
}

export default function useForm(initial = {}): Form {
  const [inputs, setInputs] = useState(initial)

  function handleChange(e: { target: { files?: any; name?: any; type?: any; value?: any } }) {
    const { name, type } = e.target
    let { value } = e.target
    if (type === 'number') {
      value = parseInt(value, 10)
    }
    if (type === 'file') {
      [value] = e.target.files
      uploadFile({ value, inputs, setInputs })
    }
    setInputs({
      ...inputs,
      [name]: value,
    })
  }

  function resetForm() {
    setInputs(initial)
  }

  function clearForm() {
    const blankState = Object.fromEntries(Object.entries(inputs).map(([key]) => [key, '']))
    setInputs(blankState)
  }

  return {
    inputs,
    setInputs,
    handleChange,
    resetForm,
    clearForm,
  }
}
