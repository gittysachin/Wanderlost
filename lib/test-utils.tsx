import React from 'react'
import { render as defaultRender, RenderResult } from '@testing-library/react'
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import { NextRouter } from 'next/router'
import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'

export * from '@testing-library/react'

import { GetItem_item } from '../components/__generated__/GetItem'
import {
  GetOrder_order,
  GetOrder_order_items,
} from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// --------------------------------------------------
// Override the default test render with our own
//
// You can override the router mock like this:
//
// const { baseElement } = render(<MyComponent />, {
//   router: { pathname: '/my-custom-pathname' },
// });
// --------------------------------------------------
type DefaultParams = Parameters<typeof defaultRender>
type RenderUI = DefaultParams[0]
type RenderOptions = DefaultParams[1] & { router?: Partial<NextRouter> }

export function render(
  ui: RenderUI,
  { wrapper, router, ...options }: RenderOptions = {}
): RenderResult {
  if (!wrapper) {
    wrapper = ({ children }) => {
      const Wrapper = (
        <RouterContext.Provider value={{ ...mockRouter, ...router }}>
          {children}
        </RouterContext.Provider>
      )
      return Wrapper
    }
  }

  return defaultRender(ui, { wrapper, ...options })
}

const mockRouter: NextRouter = {
  basePath: '',
  pathname: '/',
  route: '/',
  asPath: '/',
  query: {},
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
}

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (
  overrides: Partial<GetCurrentUser_user_cart>
): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
