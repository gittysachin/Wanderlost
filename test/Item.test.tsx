import { MockedProvider } from '@apollo/react-testing'
import { render, fakeItem } from '../lib/test-utils'
import ItemComponent from '../components/Item'

const item = fakeItem()

describe('<Item/>', () => {
  test('renders and matches the snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <ItemComponent item={item} />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })

  test('renders the image properly', () => {
    const { getByAltText } = render(
      <MockedProvider>
        <ItemComponent item={item} />
      </MockedProvider>
    )
    const img = getByAltText(item.title)
    expect(img).toBeInTheDocument()
  })

  test('renders the pricetag and title', () => {
    const { container, getByText } = render(
      <MockedProvider>
        <ItemComponent item={item} />
      </MockedProvider>
    )
    expect(getByText('$50')).toBeInTheDocument()
    const link = container.querySelector('a')
    expect(link).toHaveAttribute('href', '/item/abc123')
    expect(link).toHaveTextContent(item.title)
  })

  test('renders out the buttons properly', () => {
    const { getByText } = render(
      <MockedProvider>
        <ItemComponent item={item} />
      </MockedProvider>
    )

    const edit = getByText(/Edit/i)
    // @ts-expect-error href is not detected by typescript but test passes
    expect(edit.href).toContain('update?id=abc123')

    const addToCart = getByText(/add to cart/i)
    expect(addToCart).toHaveProperty('type', 'button')
    expect(addToCart).toBeInTheDocument()

    const deleteItem = getByText(/delete/i)
    expect(deleteItem).toHaveProperty('type', 'button')
    expect(deleteItem).toBeInTheDocument()
  })
})
