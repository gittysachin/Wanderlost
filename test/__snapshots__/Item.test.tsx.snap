// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item/> renders and matches the snapshot 1`] = `
<div>
  <div
    class="sc-gtssRu fSWVM"
  >
    <img
      alt="dogs are best"
      src="dog-small.jpg"
    />
    <h3
      class="sc-bdnylx eaHfPd"
    >
      <a
        href="/item/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="sc-dlnjPT dtvkQQ"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit ✏️
      </a>
      <button
        type="button"
      >
        Add
         To Cart 🛒
      </button>
      <button
        type="button"
      >
        Delete this Item
      </button>
    </div>
  </div>
</div>
`;
